dawn/src/dawn/Compiler/DawnCompiler.cpp

```diff
@@ -170,184 +209,203 @@ std::unique_ptr<OptimizerContext> DawnCompiler::runOptimizer(std::shared_ptr<SIR
     const Array3i domain_size{options_->domain_size_i, options_->domain_size_j,
                               options_->domain_size_k};
-    CG = std::make_unique<codegen::cuda::CudaCodeGen>(
-        optimizer->getStencilInstantiationMap(), *diagnostics_, options_->MaxHaloPoints,
-        options_->nsms, options_->maxBlocksPerSM, domain_size);
+    auto CG = std::make_unique<codegen::cuda::CudaCodeGen>(stencilInstantiationMap, *diagnostics_,
+                                                           options_->MaxHaloPoints, options_->nsms,
+                                                           options_->maxBlocksPerSM, domain_size);
+    return CG->generateCode();
   } else if(options_->Backend == "c++-opt") {
     dawn_unreachable("GTClangOptCXX not supported yet");
   } else {
     diagnostics_->report(buildDiag("-backend", options_->Backend,
                                    "backend options must be : " +
                                        dawn::RangeToString(", ", "", "")(std::vector<std::string>{
                                            "gridtools", "c++-naive", "c++-opt", "c++-naive-ico"})));
-    return nullptr;
+    throw std::runtime_error("An error occurred.");
+  }
+}
+
+std::unique_ptr<codegen::TranslationUnit>
+DawnCompiler::compile(const std::shared_ptr<SIR>& stencilIR) {
+  diagnostics_->clear();
+  diagnostics_->setFilename(stencilIR->Filename);
+
+  IIRSerializer::Format serializationKind = IIRSerializer::Format::Json;
+  if(options_->SerializeIIR || (options_->DeserializeIIR != "")) {
+    if(options_->IIRFormat == "json") {
+      serializationKind = IIRSerializer::Format::Json;
+    } else if(options_->IIRFormat == "byte") {
+      serializationKind = IIRSerializer::Format::Byte;
+    } else {
+      dawn_unreachable("Unknown SIRFormat option");
+    }
+  }
+
+  // Check if options are valid
+  // -max-halo
+  if(options_->MaxHaloPoints < 0) {
+    diagnostics_->report(buildDiag("-max-halo", options_->MaxHaloPoints,
+                                   "maximum number of allowed halo points must be >= 0"));
+    throw std::runtime_error("An error occurred.");
+  }
+
+  std::map<std::string, std::shared_ptr<iir::StencilInstantiation>> stencilInstantiationMap;
+
+  if(options_->DeserializeIIR == "") {
+    stencilInstantiationMap = parallelize(stencilIR);
+    if(!options_->Debug) {
+      stencilInstantiationMap = optimize(stencilInstantiationMap);
+    }
+  } else {
+    throw std::runtime_error("Trying to deserialize IIR.");
+    // Initialize optimizer
+    OptimizerContext::OptimizerContextOptions optimizerOptions;
+    if(options_) {
+      optimizerOptions = createOptimizerOptionsFromAllOptions(*options_);
+    }
+
+    auto optimizer =
+        std::make_unique<OptimizerContext>(getDiagnostics(), optimizerOptions, nullptr);
+    auto instantiation =
+        IIRSerializer::deserialize(options_->DeserializeIIR, optimizer.get(), serializationKind);
+    optimizer->restoreIIR("<restored>", instantiation);
```

Comment:
why does all of this happen after the throw?

Reply:
It's the code that was there and should be moved outside the class once we get a test that exercises it.